### ISSUE TITLE
git Mediawiki push failing on wiki with Kerberos Authentication

### ISSUE THREAD

**Issue by trenta3** [Git-Mediawiki/Git-Mediawiki#44]
A MediaWiki installation is set up as per https://www.mediawiki.org/wiki/Extension:LDAP_Authentication/Kerberos_Configuration_Examples under the section "Apache configuration for allowing Kerberos login without protecting an entire wiki" (with Kerberos Login only for edits, the wiki is publicly viewable).

The problem arise when trying to push to the wiki:

- not setting `mwLogin` result in error
  ```
  Fatal: Error 5 from mediawiki: Unable to get an edit token for action 'edit'`.
  ```

- Setting `mwLogin` to my username on that wiki result in error
  ```
  Failed to log in mediawiki user "<username>" on https://<sitename>/<wikipath>
  (error 4: Login Failure - WrongPass)
  ```
  after asking for password. "Manual Login" on the web succeeds.

I think the problem is related to the fact that a Kerberos Login is attempted only when a user visits the page `/<wikipath>/index.php/Kerberos_Login` and the upload script does not check for this possibility.

I feel this is really a short edit to git-mw code but I lack any perl experience and I could also not find the relevant authentication piece of code.

Thanks in advance for any help.

**Comment by moy** [comment]
I used git-mediawiki on a wiki with LDAP authentication (not Kerberos AFAIK, but I guess it doesn't matter for the web API side, only on the internal communication between MediaWiki and the authentication server). So, there's hope that it should work.

Did you try setting `remote.origin.mwDomain` to the appropriate domain (in my case, it's the name that appears in the dropdown in the login page)?

**Comment by trenta3** [comment]
Yes, I did try setting `remote.origin.mwDomain` to `<sitename>` (in uppercase letters, which is the Kerberos Realm) and it did not work.

Our configuration is, as reported in the previous link,
```
<Location /<wikipath>/index.php/Kerberos_Login>
  SSLRequireSSL
  AuthType Kerberos
  AuthName "Wiki Login"
  KrbMethodNegotiate On
  KrbMethodK5Passwd Off
  KrbLocalUserMapping On #Implemented in 5.4, strips @REALM from username
  KrbAuthRealms <realm>
  Krb5KeyTab <keytab_location>
  require valid-user
</Location>
```

I've noticed that changing the location from `/<wikipath>/index.php/Kerberos_Login` to `/<wikipath>/` also changes the error reported by git-mw:
```
Failed to log in mediawiki user "<username>" on https://<sitename>/<wikipath>
  (error 2: 401 Unauthorized : error occurred when accessing https://<sitename>/<wikipath>/api.php after 1 attempt(s))
```

The version of git-mw I'm using is commit [`7451fcdc0d3cffdb9aa79d2651830b44a8e052d6`](https://github.com/git/git/commit/7451fcdc0d3cffdb9aa79d2651830b44a8e052d6) of repository github.com/git/git.

MediaWiki version is 1.22.1. I googled and I found it is too old, so I'm trying to get it upgraded to a recent version. The issue may be related. I will report back when the wiki is upgraded.

**Comment by anarcat** [comment]
FWIW, it seems we can't login to 1.27 either, at least the test suite completely fails to do so if we update MW there, see #55. so this may not be kerberos-specific.

the test suite works with 1.21, so maybe the problem is actually with 1.22 and later!

**Comment by anarcat** [comment]
confirmed: i get login failures in Travis when working against 1.22. so it seems that the libmediawiki-perl extension has a bug!

https://travis-ci.org/Git-Mediawiki/Git-Mediawiki/jobs/306082247

**Comment by anarcat** [comment]
unconfirmed upstream reports:

https://rt.cpan.org/Public/Bug/Display.html?id=122928
https://rt.cpan.org/Public/Bug/Display.html?id=106214

here's a simple reproducer:

```perl
#!/usr/bin/perl -w

use MediaWiki::API;

my $wiki = MediaWiki::API->new;
my $remote_url = 'http://localhost:1234/wiki';
my $username = 'WikiAdmin';
my $password = 'AdminPass';
my $wiki_domain = 'localhost';

$wiki->{config}->{api_url} = "${remote_url}/api.php";

my $request = {lgname => $username,
               lgpassword => $password,
               lgdomain => $wiki_domain};
if ($wiki->login($request)) {
    print "logged in\n";
} else {
    print "login failed\n";
    print '  (error ' . $wiki->{error}->{code} . ': ' .
              $wiki->{error}->{details} . ")\n";
}
```

and here's the output here:

```sh
$ perl repro.pl 
login failed
  (error 2: Failed to decode JSON returned by http://localhost:1234/wiki/api.php
Decoding Error:
garbage after JSON object, at character offset 284 (before "<br />\n<b>Notice</b...") at /usr/share/perl5/MediaWiki/API.pm line 400.

Returned Data:
{"warnings":{"login":{"*":"Fetching a token via action=login is deprecated. Use action=query&meta=tokens&type=login instead."}},"login":{"result":"NeedToken","token":"d0809f57b218942a54276ff620b542d65a161427+\\","cookieprefix":"wikidb","sessionid":"mtpkhd9gh70s64ssondj148s37d4i521"}}<br />
<b>Notice</b>:  Uncommitted DB writes (transaction from DatabaseBase::query (SqlBagOStuff::setMulti/multi-row)). in <b>/home/anarcat/src/Git-Mediawiki/t/WEB/www/wiki/includes/db/Database.php</b> on line <b>3306</b><br />
)
```

**Comment by anarcat** [comment]
looks like this affected the Mediawiki bot as well, and this was their fix:

https://github.com/MediaWiki-Bot/MediaWiki-Bot/commit/884f07d943f4fa48fbc5c48ed8399b8b37639b13

i'm working on a patch.

**Comment by anarcat** [comment]
still no go - it looks like I can't login to *any* wiki with Mediawiki::API right now - serious limitation!

so right now we're stuck at supporting only 1.21 and earlier. it looks like the proper way to do this is with OAuth - there's a whole thing about [OAuth support](https://www.mediawiki.org/wiki/OAuth/For_Developers) since 1.27, but that requires a two-step process for each client: first a consumer key must be created in `Special:OAuthConsumerRegistration/propose` (and it must be approved!) and then a user key needs to be created by hand (because we can't seem to login by ourselves anyways).

it does not look good, I must say.

**Comment by anarcat** [comment]
the alternative is to use the new ["clientlogin" action](https://www.mediawiki.org/wiki/API:Login#The_clientlogin_action) which fakes a complete login - but then we need to handle stuff like CAPTCHA and so on. not fun.

**Comment by anarcat** [comment]
alternatively, it looks like [Mediawiki-bot](https://github.com/MediaWiki-Bot/MediaWiki-Bot) itself may be a good replacement for mediawiki::API. the API is not exactly the same, but it's similar, and it seems more active. but Mediawiki::Bot also fails a simple login:

```perl
my $bot = MediaWiki::Bot->new({
    assert      => 'bot',
    host        => 'localhost',
    operator    => 'WikiAdmin'});
$bot->{api}->{config}->{api_url} = 'http://localhost:1234/wiki/api.php';
$bot->{api}->{ua}->add_handler("request_send",  sub { shift->dump; return });
$bot->{api}->{ua}->add_handler("response_done", sub { shift->dump; return });
$bot->{api}->{ua}->default_header('Accept-Encoding', '');
$bot->login({ username => "WikiAdmin", password => "AdminPass" }) or die "login failed";

my $text = $bot->get_text("Main_Page");
print "text: $text\n";
```

**Comment by anarcat** [comment]
all this also fails with a BotPassword, FWIW.

**Comment by anarcat** [comment]
i have found a discussion about this in here, where someone is trying to fix mediawiki::API the same way, probably even to make this remote work. i have submitted an answer there, but I believe it fails as well: https://stackoverflow.com/questions/10774476/error-upload-file-using-mediawikiapi/47446018
